src/hotspot/os_cpu/linux_aarch64/vm_version_linux_aarch64.cpp

```diff
@@ -168,24 +168,43 @@ void VM_Version::get_os_cpu_info() {
   }
 }
 
-void VM_Version::get_compatible_board(char *buf, int buflen) {
+static bool read_fully(const char *fname, char *buf, size_t buflen) {
   assert(buf != NULL, "invalid argument");
   assert(buflen >= 1, "invalid argument");
-  *buf = '\0';
-  int fd = open("/proc/device-tree/compatible", O_RDONLY);
+  int fd = open(fname, O_RDONLY);
   if (fd != -1) {
-    ssize_t read_sz = read(fd, buf, buflen - 1);
-    if (read_sz > 0) {
-      buf[read_sz] = '\0';
+    ssize_t read_sz = read(fd, buf, buflen);
+    close(fd);
+
+    // Skip if the contents starts with '\n' because some machine only sets
+    // '\n' to the board name.
+    // (e.g. /sys/devices/virtual/dmi/id/board_name)
+    if ((read_sz > 0) && (*buf != '\n')) {
+      buf[read_sz - 1] = '\0';
```

Comment:
Personally, I would move it after we process the characters. It logically belongs at the end of the processing code, just before we return "true".